cilium: Fix ipip device mtu

borkmann opened the pull request:
(see commit descs)

borkmann committed:
cilium, ipip: Slightly refactor device creation

Make it a bit more readable.

Signed-off-by: Daniel Borkmann <daniel@iogearbox.net>

borkmann committed:
cilium, ipip: Fixate MTU for IPIP based on phys devs

We set up IPIP devices with no MTU specified. This is fine for a default MTU
of 1500 for phys devices on the Cilium node, however, when the MTUs are
different, then the ipip devices are still created with kernel default which
is incorrect.

Change this to adapt to the underlying MTU instead. The overhead is calculated
from the kernel defaults (delta of 1500 - default_tun_mtu):

  [...]
  47: foo@NONE: <NOARP> mtu 1452 qdisc noop state DOWN mode DEFAULT group default qlen 1000
      link/tunnel6 :: brd :: permaddr 6242:e8c9:5e84::
  48: bar@NONE: <NOARP> mtu 1480 qdisc noop state DOWN mode DEFAULT group default qlen 1000
      link/ipip 0.0.0.0 brd 0.0.0.0
  [...]

For IPv4 this is basically 20 bytes of IP header, for IPv6 tunnels this is the
overhead of IPv6 header + 8 bytes for encap limit option, see also kernel commit
381601e5bbae ("Make the ip6_tunnel reflect the true mtu.").

Signed-off-by: Daniel Borkmann <daniel@iogearbox.net>

aspsk reviewed (approved):
thanks 👍

dylandreimerink commented on pkg/datapath/loader/netlink.go:
```diff
@@ -316,17 +316,22 @@ func setupVxlanDevice(sysctl sysctl.Sysctl, port, srcPortLow, srcPortHigh uint16
-		if _, err := ensureDevice(sysctl, &netlink.Iptun{
-			LinkAttrs: netlink.LinkAttrs{Name: defaults.IPIPv4Device},
+		// sizeof(struct iphdr) for the IPIP encap.
+		ipipOverhead := 20
```
We should also reflect these numbers in https://github.com/cilium/cilium/blob/main/pkg/mtu/mtu.go#L140-L148 so that anyone relying on `RouteMTU` to make packets that should fit inside tunnels is also aware.

borkmann replied:
I'll follow-up with a comment. Right now IPIP != vxlan/geneve and we only use the former to terminate external traffic, but Pods don't send traffic via IPIP like in case of vxlan/geneve. So code-wise it's fine as-is, but I'll update the comments in a follow-up to note that down there.

asauber reviewed (changes requested):
I'm currently testing a unexpected drop scenario with this patch and running into some issues. One thing that I notice is that regardless of the host device MTU, the `cilium_tunl` device will now stay at 1480.

Will follow up as I have more details, and feel free to ping me on Cilium Slack.

nickolaev commented on pkg/datapath/loader/netlink.go:
```diff
@@ -342,11 +347,19 @@ func setupIPIPDevices(sysctl sysctl.Sysctl, ipv4, ipv6 bool) error {
-			LinkAttrs: netlink.LinkAttrs{Name: defaults.IPIPv6Device},
+		// sizeof(struct ipv6hdr) + 8 for tunnel encap limit
+		// See kernel commit 381601e5bbae ("Make the ip6_tunnel reflect the
+		// true mtu.") for details.
```
https://github.com/torvalds/linux/commit/381601e5bbae78d7c18d946fe874a63957edea13

borkmann committed:
cilium, ipip: Apply fb_tunnels_only_for_init_net before creating devs

This sysctl needs to be applied before we attempt to create the IPIP
devices since it prevents the kernel IPIP driver to create the legacy
device:

  Controls if fallback tunnels (like tunl0, gre0, gretap0, erspan0, sit0,
  ip6tnl0, ip6gre0) are automatically created when a new network namespace
  is created, if corresponding tunnel is present in initial network
  namespace. [...]

For more details, see the comment in setupIPIPDevices().

Fixes: fa94009d44b5 ("init.sh: move ipip device creation to Go")
Signed-off-by: Daniel Borkmann <daniel@iogearbox.net>

borkmann committed:
cilium, tests: Add testing for IPIP device MTU

Extend TestSetupIPIPDevices and check for different MTUs.

Signed-off-by: Daniel Borkmann <daniel@iogearbox.net>

borkmann committed:
cilium, ipip: Move IPIP overhead constants into MTU package

Move these constants into the MTU package along with the comments.

Signed-off-by: Daniel Borkmann <daniel@iogearbox.net>
